### PR TITLE
MDEV-33746 Supply missing override markings

### PR DESCRIPTION
**Pull request by DaveGosselin-MariaDB**
Find and fix missing virtual override markings.  Updates CMAKE_CXX_FLAGS to include -Wsuggest-override and -Winconsistent-missing-override.  In a few places, redundant virtual markings were removed.

### CODE REVIEW

File: sql/item_cmpfunc.h
Comment (dr-m) [inline sql/item_cmpfunc.h:307]: While we are changing the lines, we could also fix any `-Wunused-parameter` warnings:
```c++
  bool find_not_null_fields(table_map) override { return false; }
```

File: sql/item_row.h
Comment (dr-m) [inline sql/item_row.h:81]: Side note (to be handled in a separate fix): C-style casts should be avoided, and string literals are already compatible with `const char*` by default.

Some savings could be had if `Item_row::illegal_method_call(const char*)` returned a `size_t` of 0. In that case, tail call optimization could be enabled:
```c++
return illegal_method_call("val_int");
```
or
```c++
return reinterpret_cast<String*>(illegal_method_call("val_str"));
```

File: sql/item_strfunc.h
Comment (dr-m) [inline sql/item_strfunc.h:873]: `arg` is unused.

File: sql/item_subselect.h
Comment (dr-m) [inline sql/item_subselect.h:363]: I’d be consistent and remove `virtual` from anything that is declared `override`.

File: sql/log_event.h
Comment (dr-m) [inline sql/log_event.h:3022]: It would be better to indent nested preprocessor conditions with one space per nesting level after the initial `#`:
```c++
#ifdef MYSQL_SERVER
  bool write() override;
# ifdef HAVE_REPLICATION
  bool is_part_of_group() override { return true; }
# endif
```

File: sql/log_event.h
Comment (dr-m) [inline sql/log_event.h:5684]: This could be replaced with `#else`.

File: sql/sp_head.h
Comment (dr-m) [inline sql/sp_head.h:1592]: `dst_ctx` is an unused parameter and should be omitted. Also the `virtual` is redundant because it is already implied by `override`.

**Review by dr-m — APPROVED** [review]
Thank you. As far as I understand, this may omit some dead replication code when the code is built without `HAVE_REPLICATION`, I guess, as part of the embedded server library.

I have some additional suggestions:
* Ensure that `git grep 'virtual.*override'` returns no results.
* Add the warning flags to `cmake/maintainer.cmake` in order to prevent regressions.
* Try to ensure that the names of any unused parameters will be omitted from the changed definitions, to suppress `-Wunused-parameter`.

**Comment by DaveGosselin-MariaDB** [comment]
> Thank you. As far as I understand, this may omit some dead replication code when the code is built without `HAVE_REPLICATION`, I guess, as part of the embedded server library.

Especially in log_event.h, there are some methods that are marked `override` that, without the additional `#ifdef`s I added, would be marked during compilation as an error because they don't override any `virtual` methods.  For example, the `virtual is_part_of_group()` method is present only when `#if defined(MYSQL_SERVER) && defined(HAVE_REPLICATION)` at line 1478 is true.

> I have some additional suggestions:
> 
> * Ensure that `git grep 'virtual.*override'` returns no results.

Completely agree, will incorporate in this patch.

> * Add the warning flags to `cmake/maintainer.cmake` in order to prevent regressions.

Great suggestion, will do.

> * Try to ensure that the names of any unused parameters will be omitted from the changed definitions, to suppress `-Wunused-parameter`.

This is a great idea, but I'd prefer to implement it in a separate delivery, mostly because it's independent of `override` markings.

File: sql/item_cmpfunc.h
Comment (DaveGosselin-MariaDB) [reply sql/item_cmpfunc.h:307]: Agreed that these should be fixed but in a different patch.

File: sql/item_subselect.h
Comment (DaveGosselin-MariaDB) [reply sql/item_subselect.h:363]: Agreed, will do as part of this patch.

File: storage/perfschema/ha_perfschema.h
Comment (dr-m) [inline storage/perfschema/ha_perfschema.h:92]: In C++, the empty parameter list should be written as `()` and not `(void)`. This would be easy to check with the following:
```sh
git show|grep '^+.*(void)'
```

File: storage/perfschema/pfs_engine_table.cc
Comment (dr-m) [inline storage/perfschema/pfs_engine_table.cc:346]: `code` and `fmt` are unused.

File: storage/perfschema/pfs_engine_table.h
Comment (dr-m) [inline storage/perfschema/pfs_engine_table.h:344]: It looks like this should be indented by 2 columns less.

File: storage/test_sql_discovery/test_sql_discovery.cc
Comment (dr-m) [inline storage/test_sql_discovery/test_sql_discovery.cc:77]: `thd` is unused.

File: storage/test_sql_discovery/test_sql_discovery.cc
Comment (dr-m) [inline storage/test_sql_discovery/test_sql_discovery.cc:74]: All parameters are unused.

File: storage/perfschema/table_setup_instruments.h
Comment (dr-m) [inline storage/perfschema/table_setup_instruments.h:115]: Please do not add lines with trailing space.

File: storage/perfschema/table_setup_objects.h
Comment (dr-m) [inline storage/perfschema/table_setup_objects.h:87]: Trailing space should be easy to spot with
```sh
git show|grep '^+.*[	 ]$'
```

File: storage/sequence/sequence.cc
Comment (dr-m) [inline storage/sequence/sequence.cc:93]: All parameters are unused.

**Review by dr-m — APPROVED** [review]
Great work. I found some formatting nitpicks, but did not flag all of them.

I would suggest running a few more formatting checks:
```sh
git show|grep '^+.\{80\}'
git show|grep '^+.*unused'
```
You can ignore the long-line output for files that are already clearly violating it, such as `storage/sphinx`.

This does not need any additional review from me.

**Pull request merged**
